File: library/src/main/java/wannabe/grid/BoundingGrid.java
package wannabe.grid;

import java.util.Iterator;
import wannabe.Bounds;
import wannabe.Voxel;
import wannabe.grid.iterators.InBoundsIterator;

/** A Grid that removes all voxels that lie outside the specified bounds. */
public class BoundingGrid implements Grid {
  private final String name;
  private final Grid source;
  private final Bounds bounds;

  public BoundingGrid(String name, Grid source, Bounds bounds) {
    this.name = name;
    this.source = source;
    this.bounds = bounds;
  }

  @Override public Iterator<Voxel> iterator() {
    return new InBoundsIterator(source, bounds);
  }

  @Override public boolean isDirty() {
    return source.isDirty();
  }

  @Override public int size() {
    // TODO this is a lie
    return source.size();
  }

  @Override public AllNeighbors neighbors(Voxel voxel) {
    // TODO this is a lie
    return source.neighbors(voxel);
  }

  @Override public String toString() {
    return name + " no hidden (size: " + size() + ")";
  }
}

File: library/src/main/java/wannabe/grid/BoundingGrid.kt
package wannabe.grid

import wannabe.Bounds
import wannabe.Voxel
import wannabe.grid.iterators.inBounds

/** A Grid that removes all voxels that lie outside the specified bounds.  */
class BoundingGrid(
  private val name: String,
  private val source: Grid,
  private val bounds: Bounds
) : Grid {
  override fun iterator(): Iterator<Voxel> = inBounds(bounds)

  override val isDirty: Boolean
    get() = source.isDirty

  // TODO this is a lie
  override val size: Int
    get() = source.size

  // TODO this is a lie
  override fun neighbors(voxel: Voxel): AllNeighbors = source.neighbors(voxel)

  override fun toString(): String = "$name bounding (size: $size)"
}
